library/authenticators.c

```diff
@@ -0,0 +1,143 @@
+            send_error_event(wrapped_ctx, ZITI_INVALID_AUTHENTICATOR_TYPE);
+        } else if (strncmp(err->code, UNAUTHORIZED, strlen(UNAUTHORIZED)) == 0){
+            send_error_event(wrapped_ctx, ZITI_INVALID_AUTHENTICATOR_CERT);
+        } else {
+            send_error_event(wrapped_ctx, ZITI_WTF);
+        }
+
+        return;
+    } else {
+        ZTX_LOG(INFO, "certificate authenticator extension occurred for id: %s, raising event", wrapped_ctx->authenticator_id);
+        ziti_event_t ev = {
+                .type = ZitiAuthenticatorEvent,
+                .event.authenticator = {
+                        .ctx = wrapped_ctx->ctx,
+                        .authenticator_id = wrapped_ctx->authenticator_id,
+                        .error = ZITI_OK,
+                        .is_verified = false,
+                        .new_client_cert_pem = resp->client_cert_pem
+                }
+        };
+
+        ziti_send_event(wrapped_ctx->ztx, &ev);
+    }
+
+    FREE(wrapped_ctx->authenticator_id);
+    FREE(wrapped_ctx->csr_pem);
+    FREE(wrapped_ctx);
+
+    free_ziti_extend_cert_authenticator_resp(resp);
+}
+
+static void verify_cb(void* empty, const ziti_error* err, void* ctx){
+    authenticator_ctx* wrapped_ctx = (authenticator_ctx*)ctx;
+    ziti_context ztx = wrapped_ctx->ztx;
+
+    if(err) {
+        ZTX_LOG(ERROR, "error response returned when attempting to verify extended authenticator: %d %s: %s", err->http_code, err->code, err->message);
+        if(err->http_code == 404) {
+            send_error_event(wrapped_ctx, ZITI_NOT_FOUND);
+        } else if(strncmp(err->code, CAN_NOT_UPDATE_AUTHENTICATOR, strlen(CAN_NOT_UPDATE_AUTHENTICATOR)) == 0){
+            send_error_event(wrapped_ctx, ZITI_INVALID_AUTHENTICATOR_TYPE);
+        } else if (strncmp(err->code, UNAUTHORIZED, strlen(UNAUTHORIZED)) == 0){
+            send_error_event(wrapped_ctx, ZITI_INVALID_AUTHENTICATOR_CERT);
+        } else {
+            send_error_event(wrapped_ctx, ZITI_WTF);
+        }
+    } else {
+        ZTX_LOG(INFO, "certificate authenticator extension verified successfully for id: %s, raising event", wrapped_ctx->authenticator_id);
+        ziti_event_t ev = {
+                .type = ZitiAuthenticatorEvent,
+                .event.authenticator = {
+                        .ctx = wrapped_ctx->ctx,
+                        .authenticator_id = wrapped_ctx->authenticator_id,
+                        .error = ZITI_OK,
+                        .is_verified = true,
+                        .new_client_cert_pem = NULL,
+                }
+        };
+
+        ziti_send_event(wrapped_ctx->ztx, &ev);
+    }
+
+    FREE(wrapped_ctx->authenticator_id);
+    FREE(wrapped_ctx);
+}
+
+void ziti_extend_cert_authenticator(ziti_context ztx, const char* csr_pem, void *ctx) {
+    ZTX_LOG(INFO, "attempting to extend certificate authenticator id: %s", ztx->api_session->authenticator_id);
```

Comment:
all new functions should check that api_session is present and return  code (ZITI_INVALID_STATE) if not
this code will crash if ztx is in unauthenticated state

you'll need to change the signature to return `int` as it is the only way to communicate error

Reply:
Just to confirm, for example you want void -> int on the return type:

````
int ziti_extend_cert_authenticator(...);
```

And then something like:

```
int ziti_extend_cert_authenticator(...) {
    if(ztx->api_session == NULL){
       return ZITI_INVALID_STATE;
    }
    
    ...
    
    return ZITI_OK;
}
```

Reply:
we need to either return an error code or call `cb` with it (in which case the function can still be `void`)
my preference is to return the error to be consistent with other functions

Reply:
Addressed.